### DISCUSSION TITLE
How to concat multiple rows with a single row

### DISCUSSION THREAD

**Discussion by ademchenko** [linq2db/linq2db#3771] [Q&A]
I have quite a difficult query with groupings and etc. in my **Postgres** database, but let me simplify everything.
Suppose I have two tables with int values in both. And I want to Concat (i.e. UNION ALL) max value from one table with all values from another getting something like:
```
SELECT max(intValue) FROM A
UNION ALL
SELECT intValue FROM B
```
Pretty easy, isn't it?

But, of course, we will get the compilation error trying `A.Select(i => i.IntValue).Max().Concat(B.Select(i => i.IntValue))` because max over ints is just a single value in terms of queryable and cannot be used as an argument of Concat.
The attempts to solve it by `new [] { A.Select(i => i.IntValue).Max() }`  or by `B.Select(i => i.IntValue).Append(A.Select(i => i.IntValue).Max())` are compiled but give a runtime error. 
I know about the approach `A.Select(i => i.IntValue).OrderBy(i => i.IntValue).Take(1)` but I need a direct approach with max. Is it possible to achieve that somehow?

**Comment by to11mtm** [comment]
There may be a better workaround, but offhand I would suggest:

```
///Declared in an appropriate static/shim class:
[Sql.Expression("Max({0})", ServerSideOnly = true)]
public static T SelectMax<T>(T value)
{
	throw new NotImplementedException();
}

//Given this Type
public class SampleSolutionTable
{
	public int A {get;set;}
	public string Val {get;set;}
}

//Example query:
dataConnection.GetTable<SampleSolutionTable>()
  .Select(w => w.A)
  .Concat(dataConnection.GetTable<SampleSolutionTable>()
               .Select(w => SelectMax(w.A)))
```

Something like above, will produce:
```
SELECT
	[w].[A]
FROM
	[SampleSolutionTable] [w]
UNION ALL
SELECT
	Max([w_1].[A])
FROM
	[SampleSolutionTable] [w_1]
```

It should be reversible so the max can be grabbed as first item too.

**Comment by sdanyliv** [comment]
Standard way is to using GroupBy by constant:
```cs
var maxQuery = db.A.GroupBy(x => 1).Select(g => g.Max(x => x.intValue));
var otherQuery = db.B.Select(x => x. intValue);

var query = maxQuery.Concat(otherQuery);
```

**Comment by ademchenko** [comment]
Thank you, both [to11mtm](https://github.com/to11mtm) and [sdanyliv](https://github.com/sdanyliv) for valuable ideas for working around the issue. I've tried both of them and, to be honest, neither one is ideal.

The first approach ([to11mtm](https://github.com/to11mtm)) generates a very clean and natural SQL but on the C# code level, it doesn't give the correct semantics. I.e. in a proposed way it is not possible to implement the method both on the server and client side even theoretically. I wouldn't say I need the real implementation on the client at the moment, but, at least, it's always good to have the correct C# semantics. Otherwise, it's much more difficult to understand the C# code. For example, the following code for Postgres extracting the property from the JSON object is semantically correct:
```
[Sql.Expression("{0}->>{1}", PreferServerSide = true)]
public static string PropertyOfJsonObject(object jsonObject, string propertyName) => throw new NotSupportedException();
```
Despite the fact it can't be evaluated on the client its semantics is pretty clear on the invocation point:
`users.Select(u => PropertyOfJsonObject(u.Data, "Address"))`
But the call of `users.Select(u => SelectMax(u.Time))` which even changes the number of the rows in the chain is meaningless from the C# point of view.


The second approach ([sdanyliv](https://github.com/sdanyliv)) is more or less semantically clear in the C# code but linq2db doesn't do any transformations prior to the generating the SQL code, so on the side of the database I get the code of that fake groping by an external parameter "$1", which is sent along with the query. So, the SQL code becomes unclear, longer and might (probably) take longer time to execute.


[sdanyliv](https://github.com/sdanyliv) as for me while working heavily with database the issue I've asked about is pretty common. And that would be awesome to have an easy way of making unions of aggregates with other selections having semantically clear code **both** on the C# and database side. Is that possible for linq2db to provide the extension method AsEnumerable which will do nothing except getting the code compiled successfully:
`public IEnumerable<T> AsEnumerable(T element) => new T[] {element}; //Do that on the client side, and nothing with the underlying "element" query on the server side.
`
Having such a method would be very handy.

Even further, if you agree with my proposal maybe you will decide to extend the approach on the properties as well. Consider an example:
`A.Select(i => new { MaxInt = i.ReferencedItems.Select(i => i.IntValue).Concat(new [] { i.PersonalValue }).Max() }` 
In the chain I take the referenced (by FK) table int values, union it with some property of the object itself and then calculate the aggregate over all that values.
 Maybe it could be worthwhile to be able to use `i.PersonalValue.AsEnumerable()`  in that context as well.

**Reply by sdanyliv** [reply]
Looks like a regression. `GroupBy(x => 1)`, should not produce any grouping. It is known LINQ way how to return result set from several aggregation functions. Will check.

**Comment by jods4** [comment]
I believe `.Select(x => Sql.Ext.Max(x.intValue).ToValue()))` might do the trick.
